feat(linter): add unicode sets support to `no-useless-escape` rule

camchenry opened the pull request:
- part of https://github.com/oxc-project/oxc/issues/5416

Replaces the handwritten regex parsing logic with the `oxc_regular_expression` parser, which should be more accurate and enables support for unicode sets.

camchenry commented:
* **#5975** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5975?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#5974** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5974?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @camchenry and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

DonIsaac commented on crates/oxc_linter/src/rules/eslint/no_useless_escape.rs:
```diff
@@ -424,6 +627,43 @@ fn test() {
         r"var foo = /\（([^\）\（]+)\）$|\(([^\)\)]+)\)$/;",
         r#"var stringLiteralWithNextLine = "line 1\line 2";"#,
         r"var stringLiteralWithNextLine = `line 1\line 2`;",
+        r#""use\ strict";"#,
```
Can you please add some of these cases to the expected fixes tests?

camchenry replied:
Thank you for the reminder! I meant to do this before putting this up, I will add these test cases

camchenry committed:
feat(linter): add unicode sets support to `no-useless-escape` rule (#5974)

- part of https://github.com/oxc-project/oxc/issues/5416

Replaces the handwritten regex parsing logic with the `oxc_regular_expression` parser, which should be more accurate and enables support for unicode sets.
